python/ql/lib/semmle/python/frameworks/Jinja2.qll

```diff
@@ -0,0 +1,56 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `jinja2` PyPI package.
+ * See https://jinja.palletsprojects.com.
+ */
+
+private import python
+private import semmle.python.ApiGraphs
+private import semmle.python.Concepts
+private import semmle.python.frameworks.data.ModelsAsData
+
+/**
+ * INTERNAL: Do not use
+ *
+ * Provides classes modeling security-relevant aspects of the `jinja2` PyPI package.
+ * See https://jinja.palletsprojects.com.
+ */
+module Jinja2 {
+  /** A call to `jinja2.Template`. */
+  private class Jinja2TemplateConstruction extends TemplateConstruction::Range, API::CallNode {
+    Jinja2TemplateConstruction() {
+      this = API::moduleImport("jinja2").getMember("Template").getACall()
+    }
+
+    override DataFlow::Node getSourceArg() { result = this.getArg(0) }
+  }
+
+  /** Definitions for modeling jinja `Environment`s. */
+  module EnvironmentClass {
+    /** Gets a reference to the `jinja2.Environment` class. */
+    API::Node classRef() {
+      result = API::moduleImport("jinja2").getMember("Environment")
+      or
+      result = ModelOutput::getATypeNode("jinja.Environment~Subclass").getASubclass*()
+    }
+
+    /** Gets a reference to an instance of `jinja2.Environment`. */
+    private DataFlow::TypeTrackingNode instance(DataFlow::TypeTracker t) {
+      t.start() and
+      result = EnvironmentClass::classRef().getACall()
+      or
+      exists(DataFlow::TypeTracker t2 | result = instance(t2).track(t2, t))
+    }
+
+    /** Gets a reference to an instance of `jinja2.Environment`. */
+    DataFlow::Node instance() { instance(DataFlow::TypeTracker::end()).flowsTo(result) }
```

Comment:
Changed.
In what sort of situations do we want to use type tracking as opposed to API graphs?

Reply:
In general, I would suggest using API graphs wherever possible, only using type tracking when API graphs are not sufficient. One example of where we do this is in the modelling of `pathlib`, where we want to observe that the `/` operator returns a `Path` object. Since the API graph (at present) doesn't have any way to reference the step from `a` to `a / b` (say) because the latter doesn't have a node in the API graph, we have to use type tracking instead.

In the glorious future when we get API graphs à la Ruby, we should be able to do this without using type trackers. (Well, apart from the two very general type(back)trackers that the API graph calculation is built upon.)